A method for ESGF / URS authentication

laliberte opened the pull request:
This PR is a solution to #20. It uses a very different approach that avoids monkey patching. At the moment this method provides facilities for both URS NASA EARTHDATA authentication and ESGF authentication.

A variant of this PR has been actively used in [laliberte/netcdf4_pydap](https://github.com/laliberte/netcdf4_pydap) and stress tested with [laliberte/cdb-query](https://github.com/laliberte/cdb-query). The idea is to use a ``requests.Session()`` object to carry around cookies from the authentication service. It has many advantages:

1. The ``requests.Session()`` object is thread safe.
2. The ``requests.Session()`` object can be used in other applications, for example if one simply wants to ``wget`` the file using the same authentication. This is a natural workflow with the THREDDS Server of the ESGF.
3. The ``requests.Session()`` object can be easily manipulated with high-level libraries. This is illustrated by the use of the ``mechanicalsoup`` library to parse and submit authentication forms.
4. The ``requests`` library is widely used and ``requests``-based authentication methods will most likely be easier to maintain in the future (this is speculation but maintaining [laliberte/netcdf4_pydap](https://github.com/laliberte/netcdf4_pydap) has shown me that this has been mostly true).

I have run extensive testing and everything seems to work pretty well. There however remains one major issue. After authenticating with the ESGF, fetching data brings out a bug whereas the time dimension cannot be retrieved. I have highlighted this issue in a currently-skipped test in ``test_cas_esgf.py``. If this PR is kept, then I could raise an issue to solve this bug.

Any comments?

laliberte commented:
I forgot: the build fails because environment variables with usernames and passwords are not set on travis-ci.

There are four environment variables required:
USERNAME_URS, PASSWORD_URS, OPENID_ESGF, PASSWORD_ESGF.

These were set in my forked build [laliberte/pydap](https://travis-ci.org/laliberte/pydap) and the build was successful!

jameshiebert commented on src/pydap/net.py:
```diff
@@ -20,3 +23,46 @@ def raise_for_status(response):
         )
+
+
+def follow_redirect(url, application=None, session=None,
```
Could you add some exception catching around the cookie mangling in this function? I'm not completely sure of what the guarantees are that we can expect from `session.cookies.get_dict()` (i.e. *maybe* we can rely on `requests` handling). But I would hate for a client's `GET()` call to completely fall over because we made an incorrect string processing assumption.

Or... is it possible to simplify this, but not doing any dict->string conversion on the cookies? I don't think that actually using any dict/cookie_jar features (i.e. accessing by key name), so is it sufficient to just pass `req.headers['Cookie']` through the recursive calls?

laliberte replied:
I think ``get_dict()`` is safe here. My extensive (yet limited) experience with ``requests`` is that things that you expect either work or fail in very transparent ways.

I've played with the cookie setting procedure of ``webob`` and it is a pain. I'm not sure anymore but I think I tested alternatives and this one was the one that worked...

jameshiebert commented on src/pydap/net.py:
```diff
@@ -3,15 +3,18 @@
 
-def GET(url, application=None):
+
+def GET(url, application=None, session=None):
```
Can you add to the docstring a comment about what session is expected to be?

jameshiebert replied:
Passing the session object through is possibly a little verbose, but I do like it for following [the zen of Python](https://github.com/python/peps/blob/master/pep-0020.txt#L28) and it's far preferable to the mokeypatching solution (shudder).

laliberte replied:
It is verbose but the ``session`` object is carrying so much information and is so opposite to monkeypatching that I think it's worth the clutter. I think we are in agreement here.

jameshiebert replied:
Indeed!

jameshiebert commented on src/pydap/tests/test_cas_esgf.py:
```diff
@@ -0,0 +1,71 @@
+
+    def test_basic_esgf_auth(self):
+        """Set up PyDAP to use the URS request() function"""
+        session = esgf.setup_session(os.environ['OPENID_ESGF'],
```
I didn't (previously) know that TravisCI could encrypt and make environment variables secret (or at least unprinted). Cool!

laliberte replied:
They're completely secret (if you ask for secrecy). It's pretty useful!

jameshiebert replied:
Definitely useful! I think though, that I'd rather not connect my personal auth credentials with a joint/organizational account (e.g. Pydap), so could you disable the auth tests in the Travis run? After you do that and push, we should have all green lights at the end of this PR.

laliberte replied:
It seems the credentials are hidden once they are entered in travis. The only way to access these environment variable once they are saved is to trash them and then enter new ones.

If you don't want to, I personally wouldn't mind putting in mine. I think it is fairly important to constantly ensure that both the URS and ESGF portals are accessible with ``pydap``...

jameshiebert replied:
I agree with you in principle w.r.t. ensuring that both URS and ESGF portals are accessible with pydap, but the challenges here are a real pain. We *could* use our personal credentials, but Travis doesn't make environment variables accessible to untrusted builds (e.g. pull requests). So basically, if we enable this, then *every other* pull request on this repo will fail the Travis build from here on out, regardless as to whether has anything to do with auth. I think that that's too big of an impact on the rest of the collaboration and build process. A few more comments below...

laliberte replied:
I did not realize that.

For the moment, I've exclude ``auth`` tests from travis-ci. I could create a regularly updated branch that tests for authentication on my fork. That would be far from CI but at least we could ensure some sort of monitoring...

jameshiebert commented on src/pydap/tests/test_cas_urs.py:
```diff
@@ -0,0 +1,50 @@
+                                    check_url=self.url)
+
+        test_url = self.url + '.dods?SLP[0:1:0][0:1:10][0:1:10]'
+        res = requests.get(test_url, cookies=session.cookies)
```
I'm not sure that there's value in testing the same URL using *both* requests and what PyDAP uses (`WebOb`). Can you strike the first requests.get?

jameshiebert replied:
Yeah, I guess that makes sense. If that's the case, they maybe if you could just add a comment explaining the intent, and then make sure that `requests` is included in the tests_require in `setup.py`.

laliberte replied:
I've put ``requests`` in the ``cas_extra`` requirement list. Is that enough?

jameshiebert replied:
Well, I can imagine two install scenarios (that relate to this comment):

1. A user installs the Pydap client and wants to authenticate against an ESGF or URS server.
1. A developer installs Pydap and wants to run the testing suite.

I'd argue that those are two separate cases that call for two separate sets of dependencies. E.g. `nose` might be in the latter set, but not the former set. But let's say that we have a developer working on some other part of the code base who doesn't even want to run the auth tests, that entire module is going to fail the discovery phase, w/o `requests` being installed.

For that reason, I think that it's best to have `requests` listed in both cas_extras and testing_extras.

laliberte replied:
I've done exactly as you recommended. ``resquests`` is now required in both cas_extras and testing_extras.

jameshiebert commented on src/pydap/tests/test_cas_urs.py:
```diff
@@ -0,0 +1,50 @@
+else:
+    import unittest
+
+logging.basicConfig(filename='cas_urs_test.log', level=logging.DEBUG)
```
We probably don't want this in the production code, eh?

laliberte replied:
I was actually following the other PR. Removed.

jameshiebert commented on src/pydap/tests/test_cas_esgf.py:
```diff
@@ -0,0 +1,71 @@
+logging.basicConfig(filename='cas_esgf_test.log', level=logging.DEBUG)
+
+
+class TestESGF(unittest.TestCase):
```
Could we mark all of these tests in some way that makes them easy to skip? E.g. if some dev doesn't have auth credentials, or one doesn't have a network connection (e.g. while on a plane), it's nice to be able to easily omit the tests that require connectivity.

```python
from nose.plugins.attrib import attr
@attr('auth')
@attr('prod_url')
class TestESGF(unittest.TestCase):
    ...
```

laliberte replied:
Done!

laliberte commented on setup.py:
```diff
@@ -11,6 +11,8 @@
     'Jinja2',
     'docopt',
     'gunicorn',
+    'mechanicalsoup',
```
I understood your aim of streamlining your code with ``webob``. This is why I did not recommend using the ``requests.Session()`` object directly in ``net.py``. It would have been much easier but it would not have provided the nice features ``webob`` provides.

So I think that, yes, we should put ``mechanical_soup`` and ``requests` in another requirement list as it is only necessary for authentication.

jreadey commented:
Hey, I haven't tried out this PR, but was wondering, does it require some sort of user interaction (e.g. logging into a web page)?  What about command line apps where there's not a human in the loop?

laliberte commented:
No user interaction required. Note that it is tested under travis-ci so it is a guarantee that the authentication requires no user interaction.

I have actually used a variant of this for massive data subsetting on the ESGF for many months now through [github.com/laliberte/cdb_query](laliberte/cdb_query).
For example, it is stable enough that subsetting all daily sea ice data to the northern hemisphere from the ESGF while simultaneously remapping (client side) can be accomplished in several hours without user interaction. All it requires is one command line (albeit hell of a command line).

laliberte commented:
I've pushed a new version. Let me know if you would like any more modifications.

jameshiebert commented:
@laliberte First of all thanks so much for working on this and for your patience in getting the PR solid. I have a couple more comments:

I'm not sure exactly how you got the tests to pass on Travis. I've tried to run the 4 tests that you wrote on 3 different Linux systems (Ubuntu, CentOS 7 and Gentoo), and on all of them, I've had SSL verify errors on the ESGF tests.

```python
======================================================================
ERROR: test_dimension_esgf_query (pydap.tests.test_cas_esgf.TestESGF)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/storage/home/hiebert/code/git/pydap/src/pydap/tests/test_cas_esgf.py", line 47, in test_dimension_esgf_query
    dataset = open_url(self.url, session=session)
  File "/storage/home/hiebert/code/git/pydap/src/pydap/client.py", line 59, in open_url
    dataset = DAPHandler(url, application, session).dataset
  File "/storage/home/hiebert/code/git/pydap/src/pydap/handlers/dap.py", line 50, in __init__
    r = GET(ddsurl, application, session)
  File "/storage/home/hiebert/code/git/pydap/src/pydap/net.py", line 20, in GET
    return follow_redirect(url, application=application, session=session)
  File "/storage/home/hiebert/code/git/pydap/src/pydap/net.py", line 70, in follow_redirect
    cookies_dict=cookies_dict)
  File "/storage/home/hiebert/code/git/pydap/src/pydap/net.py", line 57, in follow_redirect
    res = req.get_response(application)
  File "/local_temp/hiebert/env34/lib64/python3.4/site-packages/webob/request.py", line 1299, in send
    application, catch_exc_info=False)
  File "/local_temp/hiebert/env34/lib64/python3.4/site-packages/webob/request.py", line 1263, in call_application
    app_iter = application(self.environ, start_response)
  File "/local_temp/hiebert/env34/lib64/python3.4/site-packages/webob/client.py", line 94, in __call__
    path, body, headers)
  File "/usr/lib64/python3.4/http/client.py", line 1125, in request
    self._send_request(method, url, body, headers)
  File "/usr/lib64/python3.4/http/client.py", line 1163, in _send_request
    self.endheaders(body)
  File "/usr/lib64/python3.4/http/client.py", line 1121, in endheaders
    self._send_output(message_body)
  File "/usr/lib64/python3.4/http/client.py", line 951, in _send_output
    self.send(msg)
  File "/usr/lib64/python3.4/http/client.py", line 886, in send
    self.connect()
  File "/usr/lib64/python3.4/http/client.py", line 1268, in connect
    server_hostname=server_hostname)
  File "/usr/lib64/python3.4/ssl.py", line 365, in wrap_socket
    _context=self)
  File "/usr/lib64/python3.4/ssl.py", line 583, in __init__
    self.do_handshake()
  File "/usr/lib64/python3.4/ssl.py", line 810, in do_handshake
    self._sslobj.do_handshake()
nose.proxy.SSLError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:600)
```

Have you seen this before?

When setting up the session and searching for the proper environment variables, rather than just searching the key from the dict (e.g. os.environ['USERNAME_URS']) which could raise a KeyError, could you use os.environ.get() and provide a blank string as the default? This way, if the environment variable is not set, it will at least fail more gracefully (address not found, or HTTP 401) rather than crashing with a KeyError.

Finally, in `setup.py`, `cas_extras` actually needs to be passed to the `setup()` function to be usable. See the `extras_require` parameter.

laliberte commented:
First, I've changed ``os.environ.__getitem__`` to ``os.environ.get``, as you suggested.

As for you being able to succeed with the ESGF auth, I think I know what's wrong. It is a one-time thing where a user has to register once with the dataset authority (choosing either Research of Commercial). In the latest commit, I've added a client-level check that warns the user that they have to perform one more step through the browser. It's not perfect but once set-up, it works pretty well. I've formulated it as a ``UserWarning`` that raises an exception. I feel this is the better way but your opinion would be appreciated here.

laliberte commented:
I've tweaked my last commit a bit more and wrote a test to ensure that this error message is raised to the unregistered user.

laliberte commented:
That test was failing but is now working: we can look at my [auth testing branch](https://travis-ci.org/laliberte/pydap/builds/180292725). The PR builds properly as well. I feel that all you need to do now is check on your side the authentication tests. For these, you need:

OPENID_ESGF_NO_REG -> an openid from which you have never downloaded CORDEX data.
PASSWORD_ESGF_NO_REG -> associated password

OPENID_ESGF -> an openid from which you have SUCCESFULLY downloaded CORDEX data.
PASSWORD_ESGF -> associated password

USERNAME_URS and PASSWORD_URS

Hopefully, this will work!

jameshiebert commented:
We haven't heard back from @jgallagher59701 and @lewismc yet. Do you two have any comments on this PR?

jgallagher59701 commented:
> On Dec 1, 2016, at 13:44, James Hiebert <notifications@github.com> wrote:
> 
> We haven't heard back from @jgallagher59701 <https://github.com/jgallagher59701> and @lewismc <https://github.com/lewismc> yet. Do you two have any comments on this PR?
> 
> 
This seems fine to me. Anything that will get the code working with URS that’s easy for users is good enough.

James
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/pydap/pydap/pull/28#issuecomment-264289230>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AH8OKsXu9X3akVtXpaIDsUlvIIBSYuuhks5rDzGkgaJpZM4K9Sas>.
> 

--
James Gallagher
jgallagher@opendap.org

jameshiebert commented:
OK, if it looks good to others then I'll go ahead and merge it (with two small edits). I still have concerns about the use of using an http URL in the tests (exposes credentials) and the use of non-verified SSL for the ESGF (exposes possibility of man-in-the-middle). But I can merge this to get us started, write up an issue, and then we can address it later on.

lewismc commented:
Ack thanks excellent work folks I pulled locally and verified all
functionality.

On Thu, Dec 1, 2016 at 1:04 PM James Hiebert <notifications@github.com>
wrote:

> Merged #28 <https://github.com/pydap/pydap/pull/28>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pydap/pydap/pull/28#event-879124352>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/ABHJl-ETupRKYTY5-sKeL_7_oIhucVroks5rDzZMgaJpZM4K9Sas>
> .
>

bucricket commented:
Wonderful!  Thanks for the hard work!

Mitch

laliberte commented:
Hopefully this is going to be useful for a few people!

@jameshiebert, I share your concerns about the man-in-the-middle attack possibility but, for some reason, ESGF servers tend to have outdated certificates. This means that there are no ways to communicate with these servers with``verify=True`` until they fix their certifcate problem. These issues have been ongoing for the last 5 years so I doubt it will be fixed anytime soon...
